# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: widget/cocoa/nsChildView.mm
# Commit: 14eab025027a
# Full Hash: 14eab025027a09c77de45e0d0c910b2a37eec082
# Author: Timothy Nikkel <tnikkel@gmail.com>
# Date: 2021-03-12 15:32:35
# Regressor Bug: 1696802
# File Overlap Count: 1
# Description:
#   Bug 1696802. Add code to support double tap to zoom on mac. r=botond,mstange
#   
#   We already have a pref for double tap to zoom with is already enabled by default so it's kind of awkward to add another pref.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D107389
# ==============================================================================

diff -r 759586ace266 -r 14eab025027a widget/cocoa/nsChildView.mm
--- a/widget/cocoa/nsChildView.mm	Fri Mar 12 01:46:25 2021 +0000
+++ b/widget/cocoa/nsChildView.mm	Fri Mar 12 02:10:36 2021 +0000
@@ -1939,6 +1939,12 @@
     PinchGestureInput& pinchEvent = aEvent.AsPinchGestureInput();
     WidgetWheelEvent wheelEvent = pinchEvent.ToWidgetEvent(this);
     ProcessUntransformedAPZEvent(&wheelEvent, result);
+  } else if (aEvent.mInputType == TAPGESTURE_INPUT) {
+    TapGestureInput& tapEvent = aEvent.AsTapGestureInput();
+    WidgetSimpleGestureEvent gestureEvent = tapEvent.ToWidgetEvent(this);
+    ProcessUntransformedAPZEvent(&gestureEvent, result);
+  } else {
+    MOZ_ASSERT_UNREACHABLE();
   }
 
   return result.GetStatus();
@@ -2717,13 +2723,28 @@
 
   nsAutoRetainCocoaObject kungFuDeathGrip(self);
 
-  // Setup the "double tap" event.
-  WidgetSimpleGestureEvent geckoEvent(true, eTapGesture, mGeckoChild);
-  [self convertCocoaMouseEvent:anEvent toGeckoEvent:&geckoEvent];
-  geckoEvent.mClickCount = 1;
-
-  // Send the event.
-  mGeckoChild->DispatchWindowEvent(geckoEvent);
+  if (StaticPrefs::apz_mac_enable_double_tap_zoom_touchpad_gesture()) {
+    PRIntervalTime eventIntervalTime = PR_IntervalNow();
+    TimeStamp eventTimeStamp = nsCocoaUtils::GetEventTimeStamp([anEvent timestamp]);
+
+    NSPoint locationInWindow = nsCocoaUtils::EventLocationForWindow(anEvent, [self window]);
+    ScreenPoint position =
+        ViewAs<ScreenPixel>([self convertWindowCoordinatesRoundDown:locationInWindow],
+                            PixelCastJustification::LayoutDeviceIsScreenForUntransformedEvent);
+
+    TapGestureInput event{TapGestureInput::TAPGESTURE_DOUBLE, eventIntervalTime, eventTimeStamp,
+                          RoundedToInt(position), nsCocoaUtils::ModifiersForEvent(anEvent)};
+
+    mGeckoChild->DispatchAPZInputEvent(event);
+  } else {
+    // Setup the "double tap" event.
+    WidgetSimpleGestureEvent geckoEvent(true, eTapGesture, mGeckoChild);
+    [self convertCocoaMouseEvent:anEvent toGeckoEvent:&geckoEvent];
+    geckoEvent.mClickCount = 1;
+
+    // Send the event.
+    mGeckoChild->DispatchWindowEvent(geckoEvent);
+  }
 
   // Clear the gesture state
   mGestureState = eGestureState_None;
